hugr-llvm/src/extension/prelude/array.rs

```diff
@@ -779,4 +938,191 @@ mod test {
+        // - Creates an array [1, 2, 3, ..., size]
+        // - Maps a function that increments each element by `inc`
+        // - Returns the sum of the
+
+        let int_ty = int_type(6);
+        let hugr = SimpleHugrConfig::new()
+            .with_outs(int_ty.clone())
+            .with_extensions(exec_registry())
+            .finish_with_exts(|mut builder, reg| {
+                let mut r = builder.add_load_value(ConstInt::new_u(6, 0).unwrap());
+                let new_array_args = (0..size)
+                    .map(|i| builder.add_load_value(ConstInt::new_u(6, i).unwrap()))
+                    .collect_vec();
+                let arr = builder
+                    .add_new_array(int_ty.clone(), new_array_args)
+                    .unwrap();
+
+                let mut func = builder
+                    .define_function(
+                        "foo",
+                        Signature::new(vec![int_ty.clone()], vec![int_ty.clone()])
+                            .with_extension_delta(exec_extension_set()),
+                    )
+                    .unwrap();
+                let [elem] = func.input_wires_arr();
+                let delta = func.add_load_value(ConstInt::new_u(6, inc).unwrap());
+                let out = func.add_iadd(6, elem, delta).unwrap();
+                let func_id = func.finish_with_outputs(vec![out]).unwrap();
+                let func_v = builder
+                    .load_func(func_id.handle(), &[], &exec_registry())
+                    .unwrap();
+                let scan = ArrayScan::new(
+                    int_ty.clone(),
+                    int_ty.clone(),
+                    vec![],
+                    size,
+                    exec_extension_set(),
+                );
+                let mut arr = builder
+                    .add_dataflow_op(scan, [arr, func_v])
+                    .unwrap()
+                    .out_wire(0);
+
+                for i in 0..size {
+                    let array_size = size - i;
+                    let pop_res = builder
+                        .add_array_pop_left(int_ty.clone(), array_size, arr)
+                        .unwrap();
+                    let [elem, new_arr] = builder
+                        .build_unwrap_sum(
+                            reg,
+                            1,
+                            option_type(vec![
+                                int_ty.clone(),
+                                array_type(array_size - 1, int_ty.clone()),
+                            ]),
+                            pop_res,
+                        )
+                        .unwrap();
+                    arr = new_arr;
+                    r = builder.add_iadd(6, r, elem).unwrap();
+                }
+                builder.finish_with_outputs([r]).unwrap()
+            });
+        exec_ctx.add_extensions(|cge| cge.add_default_prelude_extensions().add_int_extensions());
+        let expected: u64 = (inc..size + inc).sum();
```

Comment:
I don't think this should work for the third case? We're summing `map (n+) (0..0)`, which should be 0, but this would have `expected = 1`

Reply:
`inc..0 + inc == inc..inc` which is an empty iterator, so shouldn't it be `expected = 0`?

Reply:
Yes you're right, somehow I read it as `(inc..size).sum() = inc`